File: templates/segment-trees/2d/2d-DoubleSparse-PointChmax-RangeMaxBIT-Offline.py
# TEMPLATE BY ISHAANBUILDSTHINGS (see my github)

from bisect import bisect_left, bisect_right

NEG = float('-inf')

class OfflineSparseBIT2DMax:
    """
    Offline sparse 2D Binary Indexed Tree over a max monoid:
      - chmax:    tree[x][y] = max(tree[x][y], v)
      - queryMax: max over a corner-anchored rectangle, orientation set
                  per axis at construction via descX / descY

    ALL update coordinates must be passed to the constructor. chmax at
    an unregistered (x, y) is undefined behaviour. That requirement is
    what buys the memory: each row node stores only the y-coords that
    actually occur inside it, sorted, with a dense inner Fenwick over
    that compressed list. A point costs log(uX) slots instead of the
    online variant's log(uX) * log(uY) -- ~34x fewer entries at
    uX = uY = 1e5, in flat lists rather than dicts.

    ORIENTATION IS FIXED AT CONSTRUCTION. descX / descY reverse an axis
    internally, turning that axis's suffix query into a prefix query.
    This cannot be a per-call option: chmax writes into node ranges
    determined by the orientation, so a tree built ascending contains no
    node covering a suffix. Choose per axis up front.

        descX  descY   queryMax(x, y) covers
        -----  -----   ---------------------
        False  False   [loX, x] x [loY, y]     both prefixes
        False  True    [loX, x] x [y, hiY]     prefix, suffix
        True   False   [x, hiX] x [loY, y]     suffix, prefix
        True   True    [x, hiX] x [y, hiY]     both suffixes

    Queries are always CORNER-ANCHORED. Max has no inverse, so an
    interior rectangle cannot be obtained by diffing prefixes.

    MONOTONE UPDATES ONLY. A cell can be raised, never lowered; a
    decrease is silently ignored, not applied.

    With P = len(points), universe sizes uX, uY:
      __init__:  O(P log P + P log(uX)) time, O(P log(uX)) memory
      chmax:     O(log(uX) * log(uY))
      queryMax:  O(log(uX) * log(uY))
    """

    def __init__(self, loX, hiX, loY, hiY, points, descX=False, descY=False):
        self.loX = loX
        self.hiX = hiX
        self.loY = loY
        self.hiY = hiY
        self.descX = descX
        self.descY = descY
        uX = hiX - loX + 1
        uY = hiY - loY + 1
        self.uX = uX
        self.uY = uY

        # map to internal 1..u once, then sort by y so every bucket comes
        # out already sorted -- no per-node sort needed, just dedup on append
        mapped = []
        for x, y in points:
            mx = hiX - x + 1 if descX else x - loX + 1
            my = hiY - y + 1 if descY else y - loY + 1
            mapped.append((my, mx))
        mapped.sort()

        buckets = [None] * (uX + 2)
        for my, mx in mapped:
            i = mx
            while i <= uX:                 # register y in every row node covering x
                b = buckets[i]
                if b is None:
                    buckets[i] = [my]
                elif b[-1] != my:
                    b.append(my)
                i += i & -i

        start = [0] * (uX + 2)             # node i owns ys[start[i] : start[i+1]]
        ys = []
        for i in range(1, uX + 1):
            start[i] = len(ys)
            b = buckets[i]
            if b:
                ys.extend(b)
        start[uX + 1] = len(ys)

        self.start = start
        self.ys = ys
        self.dat = [NEG] * len(ys)

    # O(log(uX) * log(uY))
    def chmax(self, x, y, v):
        """tree[x][y] = max(tree[x][y], v). Never lowers a cell."""
        x = self.hiX - x + 1 if self.descX else x - self.loX + 1
        y = self.hiY - y + 1 if self.descY else y - self.loY + 1
        ys = self.ys
        dat = self.dat
        start = self.start
        uX = self.uX
        while x <= uX:
            a = start[x]
            b = start[x + 1]
            size = b - a
            j = bisect_left(ys, y, a, b) - a + 1      # 1-indexed within node
            while j <= size:
                p = a + j - 1
                if dat[p] < v:
                    dat[p] = v
                j += j & -j
            x += x & -x

    # O(log(uX) * log(uY))
    def queryMax(self, x, y):
        """Max over the corner rectangle bounded by x and y, per the
        descX / descY orientation. Bounds are clamped, so out-of-range
        args are safe. NEG if empty or untouched -- no caller guard."""
        x = self.hiX - x + 1 if self.descX else x - self.loX + 1
        y = self.hiY - y + 1 if self.descY else y - self.loY + 1
        if x <= 0 or y <= 0:
            return NEG
        uX = self.uX
        uY = self.uY
        if x > uX: x = uX
        if y > uY: y = uY
        ys = self.ys
        dat = self.dat
        start = self.start
        best = NEG
        while x > 0:
            a = start[x]
            b = start[x + 1]
            j = bisect_right(ys, y, a, b) - a         # count of cols <= y
            while j > 0:
                p = a + j - 1
                cur = dat[p]
                if cur > best:
                    best = cur
                j -= j & -j
            x -= x & -x
        return best

File: templates/segment-trees/2d/2d-DoubleSparse-PointChmax-RangeMaxBIT-Offline.cpp
// TEMPLATE BY ISHAANBUILDSTHINGS (see my github)
//
// USAGE
//   vector<pair<int,int>> pts;                 // every (x, y) you will ever chmax
//   for (int i = 0; i < n; i++) pts.push_back({i, nums[i]});
//   OfflineSparseBIT2DMax nxtUp(0, n - 1, lo, hi, pts);
//   OfflineSparseBIT2DMax nxtDown(0, n - 1, lo, hi, pts, false, true);
//
// Offline sparse 2D BIT over a max monoid. Same contract as the online
// variant -- corner-anchored queries only, monotone (chmax) updates only,
// orientation fixed at construction via descX / descY -- but ALL update
// coordinates must be supplied to the constructor. chmax at an
// unregistered (x, y) is undefined behaviour.
//
// That requirement is what buys the memory: each row node stores only
// the y-coords present in it, sorted, with a dense inner Fenwick over
// that compressed list. A point costs log(uX) slots, not
// log(uX) * log(uY).
//
//   ctor:      O(P log(uX) log P) time, O(P log(uX)) memory
//   chmax:     O(log(uX) * log(uY))
//   queryMax:  O(log(uX) * log(uY))
class OfflineSparseBIT2DMax {
    public:
        static constexpr long long NEG = LLONG_MIN;
    
        OfflineSparseBIT2DMax(int loX_, int hiX_, int loY_, int hiY_,
                              const vector<pair<int,int>>& points,
                              bool descX_ = false, bool descY_ = false)
            : loX(loX_), hiX(hiX_), loY(loY_), hiY(hiY_),
              descX(descX_), descY(descY_),
              uX(hiX_ - loX_ + 1), uY(hiY_ - loY_ + 1) {
    
            vector<int> cnt(uX + 2, 0);
            for (const auto& p : points) {
                int x = descX ? hiX - p.first + 1 : p.first - loX + 1;
                for (int i = x; i <= uX; i += i & -i) cnt[i]++;
            }
            start.assign(uX + 2, 0);
            for (int i = 1; i <= uX; i++) start[i + 1] = start[i] + cnt[i];
    
            vector<int> raw(start[uX + 1]);
            vector<int> cur(start.begin(), start.end());
            for (const auto& p : points) {
                int x = descX ? hiX - p.first + 1 : p.first - loX + 1;
                int y = descY ? hiY - p.second + 1 : p.second - loY + 1;
                for (int i = x; i <= uX; i += i & -i) raw[cur[i]++] = y;
            }
    
            // sort + dedup each node, compacting into a tight array
            ys.reserve(raw.size());
            vector<int> ns(uX + 2, 0);
            for (int i = 1; i <= uX; i++) {
                ns[i] = (int)ys.size();
                int a = start[i], b = start[i + 1];
                sort(raw.begin() + a, raw.begin() + b);
                int last = -1;
                for (int p = a; p < b; p++)
                    if (raw[p] != last) { ys.push_back(raw[p]); last = raw[p]; }
            }
            ns[uX + 1] = (int)ys.size();
            start.swap(ns);
            dat.assign(ys.size(), NEG);
        }
    
        // O(log(uX) * log(uY)) -- never lowers a cell
        void chmax(int x, int y, long long v) {
            x = descX ? hiX - x + 1 : x - loX + 1;
            y = descY ? hiY - y + 1 : y - loY + 1;
            for (int i = x; i <= uX; i += i & -i) {
                int a = start[i], b = start[i + 1], sz = b - a;
                int j = int(lower_bound(ys.begin() + a, ys.begin() + b, y)
                            - ys.begin() - a) + 1;      // 1-indexed within node
                for (; j <= sz; j += j & -j)
                    if (dat[a + j - 1] < v) dat[a + j - 1] = v;
            }
        }
    
        // O(log(uX) * log(uY)) -- NEG if empty or untouched
        long long queryMax(int x, int y) const {
            x = descX ? hiX - x + 1 : x - loX + 1;
            y = descY ? hiY - y + 1 : y - loY + 1;
            if (x <= 0 || y <= 0) return NEG;
            if (x > uX) x = uX;
            if (y > uY) y = uY;
            long long best = NEG;
            for (int i = x; i > 0; i -= i & -i) {
                int a = start[i], b = start[i + 1];
                int j = int(upper_bound(ys.begin() + a, ys.begin() + b, y)
                            - ys.begin() - a);          // count of cols <= y
                for (; j > 0; j -= j & -j) {
                    long long c = dat[a + j - 1];
                    if (c > best) best = c;
                }
            }
            return best;
        }
    
    private:
        int loX, hiX, loY, hiY;
        bool descX, descY;
        int uX, uY;
        vector<int> start, ys;
        vector<long long> dat;
    };